File: practica2.c
#include <stdio.h>
#include <string.h>
#include <wchar.h>
#include <wctype.h>
#include <locale.h>


void reverse_string(wchar_t *string) {
    wchar_t *start = string;
    wchar_t *end = string;

    while (*end != L'\0') {
        end++;
    }
    end--;

    while (start < end) {
        wchar_t aux = *start;
        *start = *end;
        *end = aux;
        start++;
        end--;
    }

    wprintf(L"%ls ", string);
}


int isAlphabet(wchar_t c) {
    return (iswalpha(c) || c == L' ' || 
            c == L'á' || c == L'é' || c == L'í' || c == L'ó' || c == L'ú' ||
            c == L'Á' || c == L'É' || c == L'Í' || c == L'Ó' || c == L'Ú');
}


int checklen_and_alphabet(wchar_t *string) {
    wchar_t *start = string;
    int length = 0;
    int hasAlphabet = 0;

    while (*start != L'\0') {
        if (isAlphabet(*start)) {
            hasAlphabet = 1;
        }
        length++;
        start++;
    }

    if (length == 0) {
        wprintf(L"The string is empty. Exiting....\n");
        return 1;
    }

    if (length > 100) {
        wprintf(L"The length %d of the string exceeds 100. Exiting....\n", length);
        return 1;
    }

    if (!hasAlphabet) {
        wprintf(L"The string contains no alphabetic characters. Exiting....\n");
        return 1;
    }

    return 0;
}


void charFrecuencies(wchar_t *string, int *vowelCounter, int *consonantCounter, int vowel[5]) {
    *vowelCounter = 0;
    *consonantCounter = 0;

    while (*string != L'\0') {
        wchar_t ch = towlower(*string);

        if (ch == L'a' || ch == L'á') { (*vowelCounter)++; vowel[0]++; }
        else if (ch == L'e' || ch == L'é') { (*vowelCounter)++; vowel[1]++; }
        else if (ch == L'i' || ch == L'í') { (*vowelCounter)++; vowel[2]++; }
        else if (ch == L'o' || ch == L'ó') { (*vowelCounter)++; vowel[3]++; }
        else if (ch == L'u' || ch == L'ú') { (*vowelCounter)++; vowel[4]++; }
        else if (iswalpha(ch)) { (*consonantCounter)++; }

        string++;
    }
}


wchar_t* filterSpaces(wchar_t *string) {
    wchar_t *pointer = string;
    while (*pointer != L'\0') {
        if (*pointer == L' ') {
            *pointer = L'_';
        }
        pointer++;
    }
    return string;
}


void filterNonAlphabetic(wchar_t *string) {
    wchar_t *read = string;
    wchar_t *write = string;

    while (*read != L'\0') {
        if (isAlphabet(*read)) {
            *write = *read;
            write++;
        }
        read++;
    }

    *write = L'\0'; 
}


void filterFrecuencies(int vowel[5]) {
    for (int i = 0; i < 5; i++) {
        if (vowel[i] != 0) {
            wprintf(L"%d ", vowel[i]);
        }
    }
}

int main() {
    setlocale(LC_ALL, ""); 

    wchar_t string[1000];
    int vowelCount;
    int consonantCount;
    int vowels[5] = {0};

    wprintf(L"Type your string (Max: 100 chars):\n");
    fgetws(string, sizeof(string) / sizeof(wchar_t), stdin);

    int len = checklen_and_alphabet(string);
    if (len == 1) {
        return 1;
    }

    filterNonAlphabetic(string);
    reverse_string(string);
    filterSpaces(string);
    charFrecuencies(string, &vowelCount, &consonantCount, vowels);
    filterFrecuencies(vowels);
    wprintf(L"%d %d %ls", vowelCount, consonantCount, string);

    return 0;
}

File: goFolder/practica2.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strings"
	"unicode"
)

func reverseString(runes []rune) []rune {
	reversed := make([]rune, len(runes))
	copy(reversed, runes)
	for i, j := 0, len(reversed)-1; i < j; i, j = i+1, j-1 {
		reversed[i], reversed[j] = reversed[j], reversed[i]
	}
	return reversed
}

func isAlphabet(c rune) bool {
	return unicode.IsLetter(c) || c == ' ' ||
		c == 'á' || c == 'é' || c == 'í' || c == 'ó' || c == 'ú' ||
		c == 'Á' || c == 'É' || c == 'Í' || c == 'Ó' || c == 'Ú'
}

func checkLenAndAlphabet(runes []rune) bool {
	hasAlphabet := false

	for _, r := range runes {
		if isAlphabet(r) {
			hasAlphabet = true
			break
		}
	}

	if len(runes) == 0 {
		fmt.Println("The string is empty. Exiting....")
		return false
	}

	if len(runes) > 100 {
		fmt.Printf("The length %d of the string exceeds 100. Exiting....\n", len(runes))
		return false
	}

	if !hasAlphabet {
		fmt.Println("The string contains no alphabetic characters. Exiting....")
		return false
	}

	return true
}

func charFrequencies(runes []rune) (int, int, [5]int) {
	var vowelCount, consonantCount int
	var vowels [5]int

	for _, r := range runes {
		ch := unicode.ToLower(r)
		switch ch {
		case 'a', 'á':
			vowelCount++
			vowels[0]++
		case 'e', 'é':
			vowelCount++
			vowels[1]++
		case 'i', 'í':
			vowelCount++
			vowels[2]++
		case 'o', 'ó':
			vowelCount++
			vowels[3]++
		case 'u', 'ú':
			vowelCount++
			vowels[4]++
		default:
			if unicode.IsLetter(ch) {
				consonantCount++
			}
		}
	}

	return vowelCount, consonantCount, vowels
}

func filterSpaces(runes []rune) []rune {
	for i, r := range runes {
		if r == ' ' {
			runes[i] = '_'
		}
	}
	return runes
}

func filterNonAlphabetic(runes []rune) []rune {
	var filtered []rune
	for _, r := range runes {
		if isAlphabet(r) {
			filtered = append(filtered, r)
		}
	}
	return filtered
}

func filterFrequencies(vowels [5]int) {
	for _, v := range vowels {
		if v != 0 {
			fmt.Printf("%d ", v)
		}
	}

}

func main() {
	fmt.Println("Type your string (Max: 100 chars):")
	reader := bufio.NewReader(os.Stdin)
	input, _ := reader.ReadString('\n')
	input = strings.TrimSpace(input)
	runes := []rune(input)

	if !checkLenAndAlphabet(runes) {
		return
	}

	runes = filterNonAlphabetic(runes)

	reversed := reverseString(runes)
	fmt.Printf("%s ", string(reversed)) // 1. Imprime la cadena invertida original

	vowelCount, consonantCount, vowels := charFrequencies(reversed)
	filterFrequencies(vowels) // 2. Frecuencias de vocales individuales
	fmt.Printf("%d %d ", vowelCount, consonantCount)

	reversedWithUnderscores := filterSpaces(append([]rune(nil), reversed...))
	fmt.Printf("%s\n", string(reversedWithUnderscores)) // 3. Cadena invertida con espacios como guiones bajos
}
